### PR TITLE
Basic stream post

### LINKED ISSUE #5: Author 003-a: Posts private to author
- As an author, posts I create be private to me.

### LINKED ISSUE #7: Author 004-a: Posts Private to Friends
As an author, posts I create be private to my friends.

### LINKED ISSUE #9: Author 005-a: Private to Friends of Friends
As an author, posts I create be private to friends of friends

### LINKED ISSUE #27: Author 008-a: Create Basic Stream Post
User should be able to create a basic post
title
description
content
author
categories
published date/time
content type: text/plain, markdown, image/png;base64, image/jpg;base64, image/png;base64 w/ text
privacy buttons

### PR DESCRIPTION
**Pull request by shchau**
resolves #5 
resolves #7 
resolves #9 
resolves #27 

Creating a basic post.  

Some things in this branch are strictly related to displaying a post, please ignore those. 

TODO: 
Need to take out source and origin when endpoint no longer requires them.
Implement profile pictures/linking once endpoint can take them.

NOTES:
We don't let you create a post with image AND text, it's one or the other.
Nor do we support binary files of unknown types.

### CODE REVIEW

File: frontend/src/components/ProfileBubble.jsx [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import React, { Component } from 'react';
 import 'semantic-ui-css/semantic.min.css';
-import './styles/ProfileBubble.css'
+import './styles/ProfileBubble.css';
```
Comment (shchau) [inline frontend/src/components/ProfileBubble.jsx:3]: haha epic my dude

File: backend/api/view/LoginView.py [hunk 2]
```diff
@@ -9,6 +9,7 @@ class LoginView(generics.GenericAPIView):
     permission_classes = (permissions.AllowAny,)
 
     def post(self, request, *args, **kwargs):
+        print(request, "adgadfadfasfd")
```
Comment (derrick-w) [inline backend/api/view/LoginView.py]: xd

File: frontend/src/components/CategoriesModal.jsx [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+			.then((httpResponse) => {
+				if(httpResponse.status === 200) {
+					httpResponse.json().then((results) => {	
+					this.setState({
+						options: results.map(createCategoriesItem),
+						isFetching: false,
+						});
+					})
+				}
+				else {
```
Comment (derrick-w) [inline frontend/src/components/CategoriesModal.jsx:48]: xd dunno if you want this in master

**Review by derrick-w — CHANGES_REQUESTED** [review]

**Comment by hiufungk** [comment]
> Some things in this branch are strictly related to displaying a post, please ignore those.

instead of ignoring them, do not pull request the unrelated files

**Comment by hiufungk** [comment]
> Still need to handle source and origin, having an actual profile picture in the modal, and application/base64, and possibly some other stuff that I'm forgetting.

add todo's to the mark down what needs to be done in the future

File: frontend/src/actions/PostActions.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
+import HTTPFetchUtil from "../util/HTTPFetchUtil";
+
+export const sendPost = (urlPath, requireAuth, body) => {
+
+	console.log(body);
```
Comment (hiufungk) [inline frontend/src/actions/PostActions.js]: remove this

File: frontend/src/actions/PostActions.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    return (dispatch, getstate) => {
+        HTTPFetchUtil.sendPostRequest(urlPath, requireAuth, body)
+            .then((httpResponse) => {
+                console.log(httpResponse);
```
Comment (hiufungk) [inline frontend/src/actions/PostActions.js]: remove this

File: frontend/src/components/AnimatedButton.jsx [hunk 6]
```diff
@@ -10,7 +10,7 @@ class AnimatedButton extends Component {
 
 	render() {
 		return (
-			<div className="ui basic animated circular button removeBorder" tabIndex="0">
+			<div className="ui basic animated circular button removeBorder" tabIndex="0" onClick={this.props.clickFunction}>
```
Comment (hiufungk) [inline frontend/src/components/AnimatedButton.jsx:14]: add prop-types to components that depends on props as input

File: frontend/src/components/CategoriesModal.jsx [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+							isFetching: false,
+							});
+				}
+			})
+			.catch((error) => {
+				console.error(error, "ERROR");
+			});
+	}
+
+	handleAddition = (e, { value }) => {
```
Comment (hiufungk) [inline frontend/src/components/CategoriesModal.jsx]: change this to more descriptive name

File: frontend/src/components/CreatePostModal.jsx [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+const mapDispatchToProps = dispatch => {
+    return {
+        sendPost: (urlPath, requireAuth, requestBody) => {
+            return dispatch(PostActions.sendPost(urlPath, requireAuth, requestBody));
+        }
+    }
+}
+
+
+export default connect(mapStateToProps, mapDispatchToProps)(CreatePostModal);
```
Comment (hiufungk) [inline frontend/src/components/CreatePostModal.jsx:309]: add prop types

File: frontend/src/components/VisibilitySettings.jsx [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+			urlPath = '/api/author/' + UUID;
+			HTTPFetchUtil.getRequest(urlPath, requireAuth)
+			.then((httpResponse) => {
+				if(httpResponse.status === 200) {
+					httpResponse.json().then((results) => {
+						this.setState({
+							options: results.friends.map(createFriendItem), 
+						});
+					})
+				}
```
Comment (hiufungk) [inline frontend/src/components/VisibilitySettings.jsx:45]: add in error handling for non 200 response and display it

File: frontend/src/components/VisibilitySettings.jsx [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+			})
+			.catch((error) => {
+			console.error(error);
+			});
+	}
+	
+	
+	
+	openCloseDropdown = () => {
+		if (this.state.showModal === false) {
```
Comment (hiufungk) [inline frontend/src/components/VisibilitySettings.jsx:58]: do !this.state.showModal instead？

File: frontend/src/components/CreatePostModal.jsx [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+		else {
+			alert("Please ensure you have a title, description, categories, and content.");	
+		}	
+	}
+
+
+	render() {
+		
+		let {imagePreviewUrl} = this.state;
+		let $imagePreview = null;
```
Comment (hiufungk) [inline frontend/src/components/CreatePostModal.jsx:184]: why the "$"

File: frontend/src/components/CreatePostModal.jsx [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+		return true;
+	}
+
+	handleSubmit(event) {
+		const requireAuth = true,
+			urlPath = "/api/posts/",
+			requestBody = {
+				title: this.state.title,
+				source: "http://localhost:8000",
+				origin: "http://localhost:8000",
```
Comment (hiufungk) [inline frontend/src/components/CreatePostModal.jsx:147]: add a todo indicating this has to change in the future

**Review by hiufungk — CHANGES_REQUESTED** [review]

File: frontend/src/components/CreatePostModal.jsx (hunk 11 above)
Comment (shchau) [reply frontend/src/components/CreatePostModal.jsx:184]: Black magic.

**Comment by shchau** [comment]
> > Some things in this branch are strictly related to displaying a post, please ignore those.
> 
> instead of ignoring them, do not pull request the unrelated files

No they make me feel the good and are harmless.

**Review by derrick-w — APPROVED** [review]

**Review by hiufungk — APPROVED** [review]

File: frontend/src/components/ProfileBubble.jsx [hunk 13]
```diff
@@ -28,4 +32,14 @@ class ProfileBubble extends Component {
 	}
 }
 
+ProfileBubble.defaultProps = {
+	profilePicture: "$No profile picture provided"
```
Comment (forgeno) [inline frontend/src/components/ProfileBubble.jsx:37]: Please add comment why there is a "$" sign

**Review by forgeno — CHANGES_REQUESTED** [review]

**Review by forgeno — APPROVED** [review]

**Pull request merged**